test/prow/framework/testflow.py

```diff
@@ -0,0 +1,47 @@
+import os
+import sys
+import time
+
+project_root = os.path.dirname(os.path.dirname(__file__))
+sys.path.insert(0, project_root)
+
+from framework.exception import TestError
+
+
+class TestFlow():
+    def __init__(self, case):
+        self.case = case
+        self.steps = []
+
+    def append_step(self, step):
+        self.steps.append(step)
+
+    def run(self):
+        print("> Testcase \"{}\" started <".format(self.case))
+        undergoing_step = 0
+        total_steps = len(self.steps)
+        failed = False
+        try:
+            while undergoing_step < total_steps:
+                self.steps[undergoing_step].go_forth()
+                print("OK {}".format(self.steps[undergoing_step].get_step_name()))
+                time.sleep(3)
+                undergoing_step += 1
+        except TestError as e:
+            failed = True
+            print("FAIL {}".format(self.steps[undergoing_step].get_step_name()))
+            msg = e.msg
+            raise Exception("> Testcase \"{}\" failed at Step \"{}\": {}".format(self.case, self.steps[undergoing_step].get_step_name(), msg))
+        except Exception as e:
+            failed = True
+            print("FAIL {}".format(self.steps[undergoing_step].get_step_name()))
+            raise e
+        finally:
+            if undergoing_step >= total_steps:
+                undergoing_step = total_steps - 1
```

Comment:
if this pr has passed all the total_steps, waht use of "undergoing_step = total_steps - 1"?

Reply:
When the PR passed all the total_steps, the `undergoing_step` overflows the total length of `self.steps` array. So it's necessary to set `undergoing_step = total_steps - 1` to call `back_fn()` from the last step.